ADD GrantConditionOnAttack Trait

forcecore opened the pull request:
Useful for making Gattling weapons (Yuri's Revenge and C&C Generals) or Boris (RA2) / Natasha (RA3) call for airstrike.

Demo: https://www.youtube.com/watch?v=fQn0rspPH3s

forcecore committed:
ADD GrantConditionOnAttack Trait

Useful for making Gattling weapons (Yuri's Revenge and C&C Generals) or
Boris (RA2) / Natasha (RA3) call for airstrike.

abcdefg30 reviewed (dismissed):
Looks good to me otherwise.

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantStackableConditionOnFire.cs:
```diff
@@ -0,0 +1,172 @@
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class GrantStackableConditionOnFireInfo : ITraitInfo
```
All conditions are stackable, so I think we should come up with a clearer name for this.

forcecore replied:
It was just GrantConditionOnFire before a polishment. Will that be fine?

pchote replied:
GrantConditionOnAttack would be more consistent with the terminology used elsewhere

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAttack.cs:
```diff
@@ -0,0 +1,172 @@
+
+		[Desc("Shots required to apply an instance of the condition. If there are more instances of the condition granted than values listed,",
+			"the last value is used for all following instances beyond the defined range.")]
+		public readonly int[] RequiredShotsPerStack = { 3 };
```
Could we call this`ShotsPerInstance`?  I'd like to stay away from introducing new terminology for conditions.

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAttack.cs:
```diff
@@ -0,0 +1,172 @@
+			manager = self.Trait<ConditionManager>();
+		}
+
+		void StackCondition(Actor self, string cond)
```
Nit: Please rename these to `GrantInstance` and `RevokeInstance` for consistency.

pchote commented:
Can you please provide a simple testcase for this on top of one of the default mods?

GraionDilach commented:
I can vouch that this works wonderfully in AS, been using it for Black Lotus's building-capture for weeks now.

forcecore commented:
I've added an example mod. JEEP will shoot m60 --> Vulcan --> ChainGun.

forcecore commented:
Renamed some parameters in Info.

reaperrr commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAttack.cs:
```diff
@@ -0,0 +1,172 @@
+
+		[Desc("Shots required to apply an instance of the condition. If there are more instances of the condition granted than values listed,",
+			"the last value is used for all following instances beyond the defined range.")]
+		public readonly int[] RequiredShotsPerInstance = { 3 };
```
This default value looks a bit arbitrary, I'd prefer a default of 1 because the main (or even only) usage in the official mods will likely be to grant a single-instance condition with the first shot.

reaperrr commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAttack.cs:
```diff
@@ -0,0 +1,172 @@
+		public readonly int[] RequiredShotsPerInstance = { 3 };
+
+		[Desc("Maximum instances of the stackable condition.")]
+		public readonly int MaximumInstances = 2;
```
Likewise here.

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAttack.cs:
```diff
@@ -0,0 +1,172 @@
+			"the last value is used for all following instances beyond the defined range.")]
+		public readonly int[] RequiredShotsPerInstance = { 3 };
+
+		[Desc("Maximum instances of the stackable condition.")]
```
You missed a "stackable" here.  Change to "Maximum instances of the condition to grant".

reaperrr commented:
Looks good to me otherwise. I think you can remove the test commit, but maybe you could still keep and link here a separate branch with the testcase, if anyone else wants to test this before (or after) this is merged.

forcecore commented:
Updated. Rebased and example mod removed.

If you need the example gattling JEEP mod, unzip this:
[gatt.zip](https://github.com/OpenRA/OpenRA/files/1279384/gatt.zip)

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAttack.cs:
```diff
@@ -0,0 +1,172 @@
+
+		[WeaponReference]
+		[Desc("Name of the armament that grants this condition.")]
+		public readonly string ArmamentName = null;
```
Can we please change this to
```csharp
[Desc("Name of the armaments that grant this condition.")]
public readonly HashSet<string> ArmamentNames = new HashSet<string>();
```

The `!=` below then becomes a `! .Contains()`

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAttack.cs:
```diff
@@ -0,0 +1,172 @@
+		[Desc("The condition type to grant.")]
+		public readonly string Condition = null;
+
+		[WeaponReference]
```
`WeaponReference` is incorrect here.  These are referencing armament names, not weapons.

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAttack.cs:
```diff
@@ -0,0 +1,172 @@
+		int cooldown = 0;
+		int shotsFired = 0;
+		ConditionManager manager;
+		Target lastTarget = Target.Invalid;
```
Please add a comment that this is only set if `RevokeOnNewTarget` is true.

pchote reviewed (commented):
I haven't tested this yet, but the basic approach looks good.  Just a couple of minor code nits and then I anticipate being able to :+1: and merge.

forcecore commented:
Fixed and rebased

Mailaender commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAttack.cs:
```diff
@@ -0,0 +1,171 @@
+
+		void INotifyCreated.Created(Actor self)
+		{
+			manager = self.Trait<ConditionManager>();
```
`TraitOrDefault` otherwise the null checks are pointless or this is missing a `Requires<ConditionManagerInfo>`.

forcecore replied:
I added Requires<ConditionManagerInfo> as this module is completely useless without the condition manager.

pchote replied:
As mentioned above, that's not a great option.  Can you please make this `TraitOrDefault` and put the null checks back in so this matches the behaviour of the other traits?

Mailaender commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAttack.cs:
```diff
@@ -0,0 +1,171 @@
+		public readonly string Condition = null;
+
+		[Desc("Name of the armaments that grant this condition.")]
+		public readonly HashSet<string> ArmamentNames = new HashSet<string>();
```
Maybe prefill this with `primary` which is the default name for `Armament`.

Mailaender commented:
This does work. :white_check_mark:  I rebased my https://github.com/OpenRA/OpenRA/pull/13946 for a convenient test case.

forcecore commented:
The default armament name is primary now.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAttack.cs:
```diff
@@ -0,0 +1,171 @@
+	{
+		[GrantedConditionReference]
+		[Desc("The condition type to grant.")]
+		public readonly string Condition = null;
```
We could also add a `FieldLoader.Require` here.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAttack.cs:
```diff
@@ -0,0 +1,171 @@
+		int cooldown = 0;
+		int shotsFired = 0;
+		ConditionManager manager;
+		Target lastTarget = Target.Invalid; // only tracked when RevokeOnNewTarget is true.
```
Comments start with a capital letter on their own line please:
```

// Only tracked when RevokeOnNewTarget is true.
Target lastTarget = Target.Invalid;
```

abcdefg30 commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAttack.cs:
```diff
@@ -0,0 +1,171 @@
+
+		void GrantInstance(Actor self, string cond)
+		{
+			if (manager == null)
```
This can be removed now.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAttack.cs:
```diff
@@ -0,0 +1,171 @@
+		{
+			shotsFired = 0;
+
+			if (manager == null)
```
This as well.

forcecore commented:
Done more fixes

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAttack.cs:
```diff
@@ -0,0 +1,168 @@
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class GrantConditionOnAttackInfo : ITraitInfo, Requires<ConditionManagerInfo>
```
Our convention is to omit `Requires<ConditionManagerInfo>` from condition granting traits because that introduces to a circular dependency when the trait is conditional.  Can you please remove this for consistency?

forcecore commented:
TraitOrDefault for ConditionManager and null check

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/4d3ce47dfb8812c643b779531cde1fe5bdd19746...b656ee1bfa4e3a5c0c8a9ee9ade8325d0d949ec5)
